File: ts-sdk/src/cache/json.rpc.cache.ts
import { toQuantity } from "ethers";
import { LRUCache } from "lru-cache";
import { JsonRpcBlock } from "../blocks";
import { Logger, WithRetry } from "../utils";
import { RetryOptions } from "../utils/with.retry";
import { JsonRpcLog } from "../logs";
import { JsonRpcTransactionReceipt } from "../transactions";
import { MetricsHelper } from "../metrics";
import { Trace } from "../traces";
import { Cache } from "./cache";
import { GetJsonRpcCacheProvider } from "./get.json.rpc.cache.provider";
import { IsCacheHealthy } from "./is.cache.healthy";

export class JsonRpcCache implements Cache {
  private readonly ignoredBlocks = new LRUCache({ max: 10_000 });
  private readonly ignoredChains: { [chainId: number]: number } = {};
  private readonly CHAIN_IGNORE_DURATION_MS = 5 * 60 * 1000; // 5 minutes

  constructor(
    private readonly getJsonRpcCacheProvider: GetJsonRpcCacheProvider,
    private readonly jsonRpcCacheRetryOptions: RetryOptions,
    private readonly isCacheHealthy: IsCacheHealthy,
    private readonly metricsHelper: MetricsHelper,
    private readonly withRetry: WithRetry,
    private readonly logger: Logger
  ) {}

  private ignoreBlock(chainId: number, blockNumber: number): void {
    this.ignoredBlocks.set(`${chainId}-${blockNumber}`, true);
  }

  private isBlockIgnored(chainId: number, blockNumber: number): boolean {
    return this.ignoredBlocks.has(`${chainId}-${blockNumber}`);
  }

  private ignoreChain(chainId: number): void {
    this.ignoredChains[chainId] = Date.now() + this.CHAIN_IGNORE_DURATION_MS;
  }

  private isChainIgnored(chainId: number): boolean {
    return (
      chainId in this.ignoredChains && Date.now() < this.ignoredChains[chainId]
    );
  }

  async getLatestBlockNumber(chainId: number): Promise<string | undefined> {
    if (this.isChainIgnored(chainId)) return undefined;

    try {
      const blockNumber = await this.makeRequest(
        chainId,
        "eth_blockNumber",
        []
      );
      this.logger.debug(
        `chain ${chainId} latest cached block number: ${parseInt(blockNumber)}`
      );
      return blockNumber;
    } catch (e) {}
    return undefined;
  }

  async getBlockWithTransactions(
    chainId: number,
    blockNumber: number
  ): Promise<JsonRpcBlock | undefined> {
    if (this.isChainIgnored(chainId)) return undefined;

    try {
      return this.makeRequest(chainId, "eth_getBlockByNumber", [
        toQuantity(blockNumber),
        true,
      ]);
    } catch (e) {
      // if the block was not in cache, ignore the block so we dont try to fetch its logs or traces from cache
      this.ignoreBlock(chainId, blockNumber);
      // if last 3 blocks were not in cache, ignore the chain for a while
      if (
        this.isBlockIgnored(chainId, blockNumber - 1) &&
        this.isBlockIgnored(chainId, blockNumber - 2)
      ) {
        this.ignoreChain(chainId);
      }
    }
    return undefined;
  }

  async getLogsForBlock(
    chainId: number,
    blockNumber: number
  ): Promise<JsonRpcLog[] | undefined> {
    if (this.isChainIgnored(chainId)) return undefined;
    if (this.isBlockIgnored(chainId, blockNumber)) return undefined;

    try {
      const blockNumberHex = `0x${blockNumber.toString(16)}`;
      return this.makeRequest(chainId, "eth_getLogs", [
        { fromBlock: blockNumberHex, toBlock: blockNumberHex },
      ]);
    } catch (e) {}
    return undefined;
  }

  async getTraceData(
    chainId: number,
    blockNumber: number
  ): Promise<Trace[] | undefined> {
    if (this.isChainIgnored(chainId)) return undefined;
    if (this.isBlockIgnored(chainId, blockNumber)) return undefined;

    try {
      return this.makeRequest(chainId, "trace_block", [
        `0x${blockNumber.toString(16)}`,
      ]);
    } catch (e) {}
    return undefined;
  }

  private async makeRequest(chainId: number, methodName: string, args: any[]) {
    const isCacheHealthy = await this.isCacheHealthy(chainId);
    this.logger.debug(`isCacheHealthy(${chainId})=${isCacheHealthy}`);
    if (!isCacheHealthy) return undefined;

    const provider = await this.getJsonRpcCacheProvider(chainId);
    const requestId = this.metricsHelper.startJsonRpcCacheTimer(
      chainId,
      methodName
    );
    try {
      const value = await this.withRetry(
        provider.send.bind(provider),
        [methodName, args],
        this.jsonRpcCacheRetryOptions
      );
      this.metricsHelper.reportJsonRpcCacheSuccess(
        requestId,
        chainId,
        methodName
      );
      return value;
    } catch (e) {
      this.logger.debug(
        `gave up fetching ${methodName} from rpc cache for chain ${chainId}: ${e}`
      );
      this.metricsHelper.reportJsonRpcCacheError(
        requestId,
        chainId,
        methodName
      );
      throw e; // i.e. cache miss
    }
  }

  async setBlockWithTransactions(chainId: number, block: JsonRpcBlock) {
    // no-op
  }

  async setLogsForBlock(
    chainId: number,
    blockNumber: number,
    logs: JsonRpcLog[]
  ) {
    //no-op
  }

  async setTraceData(
    chainId: number,
    blockNumberOrTxHash: string | number,
    traces: Trace[]
  ) {
    // no-op
  }

  async getTransactionReceipt(
    chainId: number,
    txHash: string
  ): Promise<JsonRpcTransactionReceipt | undefined> {
    return undefined; // no-op
  }

  async setTransactionReceipt(
    chainId: number,
    txHash: string,
    receipt: JsonRpcTransactionReceipt
  ) {
    // no-op
  }

  async getAlert(alertHash: string): Promise<object | undefined> {
    return undefined; //no-op
  }

  async setAlert(alertHash: string, alert: object): Promise<void> {
    // no-op
  }

  async dump() {
    // no-op
  }
}

File: py-sdk/src/forta_bot_sdk/cache/json_rpc_cache.py
from typing import Any
from pylru import lrucache
from ..utils import Logger, WithRetry, RetryOptions, now, format_exception
from ..metrics import MetricsHelper
from .get_json_rpc_cache_provider import GetJsonRpcCacheProvider
from .is_cache_healthy import IsCacheHealthy
from .cache import Cache

CHAIN_IGNORE_DURATION_SECONDS = 5*60  # 5 minutes


class JsonRpcCache(Cache):
    def __init__(self,
                 get_json_rpc_cache_provider: GetJsonRpcCacheProvider,
                 json_rpc_cache_retry_options: RetryOptions,
                 is_cache_healthy: IsCacheHealthy,
                 metrics_helper: MetricsHelper,
                 with_retry: WithRetry,
                 logger: Logger):
        self.get_json_rpc_cache_provider = get_json_rpc_cache_provider
        self.json_rpc_cache_retry_options = json_rpc_cache_retry_options
        self.is_cache_healthy = is_cache_healthy
        self.metrics_helper = metrics_helper
        self.with_retry = with_retry
        self.logger = logger
        self.ignored_blocks = lrucache(10_000)
        self.ignored_chains: dict[int, int] = {}

    def ignore_block(self, chain_id: int, block_number: int):
        self.ignored_blocks[f'{chain_id}-{block_number}'] = True

    def is_block_ignored(self, chain_id: int, block_number: int) -> bool:
        return f'{chain_id}-{block_number}' in self.ignored_blocks

    def ignore_chain(self, chain_id: int):
        self.ignored_chains[chain_id] = now() + CHAIN_IGNORE_DURATION_SECONDS

    def is_chain_ignored(self, chain_id: int):
        return chain_id in self.ignored_chains and now() < self.ignored_chains[chain_id]

    async def get_latest_block_number(self, chain_id: int) -> str | None:
        if self.is_chain_ignored(chain_id):
            return None

        try:
            block_number_hex = await self.make_request(chain_id, "eth_blockNumber", [])
            return block_number_hex
        except Exception as e:
            self.logger.debug(
                f'get_latest_block_number:error: {format_exception(e)}')
            return None

    async def get_block_with_transactions(self, chain_id: int, block_number: int) -> dict | None:
        if self.is_chain_ignored(chain_id):
            return None

        try:
            return await self.make_request(chain_id, "eth_getBlockByNumber", [hex(block_number), True])
        except:
            # if the block was not in cache, ignore the block so we dont try to fetch its logs or traces from cache
            self.ignore_block(chain_id, block_number)
            # if last 3 blocks were not in cache, ignore the chain for a while
            if self.is_block_ignored(chain_id, block_number-1) and self.is_block_ignored(chain_id, block_number-2):
                self.ignore_chain(chain_id)
        return None

    async def get_logs_for_block(self, chain_id: int, block_number: int) -> list[dict] | None:
        if self.is_chain_ignored(chain_id):
            return None
        if self.is_block_ignored(chain_id, block_number):
            return None

        try:
            block_number_hex = hex(block_number)
            return await self.make_request(chain_id, "eth_getLogs", [{'fromBlock': block_number_hex, 'toBlock': block_number_hex}])
        except:
            return None

    async def get_trace_data(self, chain_id: int, block_number: int) -> list[dict] | None:
        if self.is_chain_ignored(chain_id):
            return None
        if self.is_block_ignored(chain_id, block_number):
            return None

        try:
            return await self.make_request(chain_id, "trace_block", [hex(block_number)])
        except:
            return None

    async def make_request(self, chain_id: int, method_name: str, args: list[Any]):
        is_cache_healthy = await self.is_cache_healthy(chain_id)
        self.logger.debug(f'isCacheHealthy({chain_id})={is_cache_healthy}')
        if not is_cache_healthy:
            return None

        provider = self.get_json_rpc_cache_provider(chain_id)
        request_id = self.metrics_helper.start_json_rpc_cache_timer(
            chain_id, method_name)
        try:
            # self.logger.debug(
            #     f'making json_rpc_cache request for {method_name} (now={now()})')
            response = await self.with_retry(provider.provider.make_request, method_name, args, retry_options=self.json_rpc_cache_retry_options)
            # self.logger.debug(
            #     f'completed json_rpc_cache request for {method_name} (now={now()})')
            self.metrics_helper.report_json_rpc_cache_success(
                request_id, chain_id, method_name)
            return response['result']
        except Exception as e:
            self.logger.debug(
                f'gave up fetching {method_name} from rpc cache for chain {chain_id}: {e}')
            self.metrics_helper.report_json_rpc_cache_error(
                request_id, chain_id, method_name)
            raise e
